scanner/image/scanner/konflux.Dockerfile

```diff
@@ -0,0 +1,71 @@
+ARG MAPPINGS_REGISTRY=registry.access.redhat.com
+ARG MAPPINGS_BASE_IMAGE=ubi8
+ARG MAPPINGS_BASE_TAG=latest
+ARG BASE_REGISTRY=registry.access.redhat.com
+ARG BASE_IMAGE=ubi8-minimal
+ARG BASE_TAG=latest
+
+FROM brew.registry.redhat.io/rh-osbs/openshift-golang-builder:rhel_8_1.20 as builder
+
+ENV CGO_ENABLED=1
+ENV GOFLAGS=""
+ENV GOTAGS="release"
+ENV RELEASE_TAG="$CI_VERSION"
```

Comment:
https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/merge_requests/2745 changed the variable from `RELEASE_TAG` to `BUILD_TAG` which has long history but is a simple change eventually.

The problem is that `CI_VERSION` is not available in Konflux context. We need to use something meaningful here and this warrants a discussion and a decision with the Scanner team.

I'd suggest you set BUILD_TAG to 1.0.0 and create a JIRA with a `# TODO(ROX-....): ...` comment here to follow up.

Reply:
If we hard coded BUILD_TAG to 1.0.0, wouldn't we always override the version string?  I thought we'd want to use the output of `make tag` for dev/CI builds.

Reply:
Yes, we need to solve `make tag`. I thought this could be done in a separate ticket, in a separate PR. But, if you want to make it here - absolutely.
Yet, `$CI_VERSION` is not set in Konflux and I suspect will be empty.

Scanner V4 adds its own vulns update semantics on top of `make tag` thing which we touched in the MR I linked.

Reply:
My thinking is that we want to handle the upstream case for now and leave the downstream for later (i.e. create a Jira and a TODO comment).  The upstream case would be to use `make tag` as-is since that gives a more descriptive version string when wanting to know where a particular image came from, for example.

Looking at the [code](https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L198), the version that gets baked into the binary is not actually used to download vulnerabilities from Central, so there should be no issue using `make tag` in regards to downloading vulnerability updates (for now at least).

Perhaps the Jira should include addressing the hard-coded replacement for ROX_VERSION in the vulnerability URL?

Reply:
> Perhaps the Jira should include addressing the hard-coded replacement for ROX_VERSION in the vulnerability URL?

Mutating side-effects in a function called `validate()` worry me also.

----

> when I deployed a Konflux-built image of scanner, dev was still used as the version string

Maybe that's because the overriding environment variable is still `RELEASE_TAG` and not `BUILD_TAG`?

That's how I trace the workings but maybe I'm wrong because I did not run the code
1. https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L199-L201 will execute if we have `ReleaseBuild`
2. We should have `ReleaseBuild` because of `ENV GOTAGS="release"` here https://github.com/stackrox/stackrox/pull/9864/files#diff-3fe9bb693b9137ea1603abd2ed863434d2e11f55b3ca1fff59d4f06deaeb4b74R12
3. The magic `Version` https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/internal/version/version.go#L1-L4 is set in the Makefile https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/Makefile#L31-L34 which will set it to whatever comes in `$(TAG)`. That, in turn, delegates to the Makefile in the repo root. https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/scanner/Makefile#L11-L13 The relevant part in the root Makefile is https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/Makefile#L29-L31 Notice that it does `git describe ...` only when `TAG` is not set. Which will be set in the imported `env.mk` with the value that comes in `BUILD_TAG` https://github.com/stackrox/stackrox/blob/b1f12088bbd991021dd692b8e7948bff6b6eed12/make/env.mk#L52-L54
4. In the current `konflux.Dockerfile` `BUILD_TAG` is not set and so `Version` should be a result of `git describe`, that `4.4.x-178-gbc0362419f` you showed. However, what escapes me is why `ENV GOTAGS="release"` does not have effect (in step (2)).

I don't know if there's any condition when this block gets executed https://github.com/stackrox/stackrox/blob/ac7ddc35c638111ce5ff7609fed986a2310d3308/scanner/config/config.go#L202-L204

What happens further in Central with `/api/extensions/scannerdefinitions?version=<ROX_VERSION>` is also interesting although I could not quite parse how the version is applied in https://github.com/stackrox/stackrox/blob/master/central/scannerdefinitions/handler/handler.go

Overall, I feel product version and vulnerability version are unnecessarily entangled which can hardly be beneficial.

I guess, I'd leave it to you to decide about the next step since you're closer to this code, can test and see effects.
The only certain thing form at this point is that `ENV RELEASE_TAG="$CI_VERSION"` has no effect on anything and should be removed.

Reply:
I found that the vuln update bundle should be copied to a `1.0.0` version: 

https://github.com/stackrox/stackrox/blob/7a197936f8d9add99e5aef85ded1ba822ff6706c/.github/workflows/scanner-dev-vuln-update.yaml#L46

But when I tried to download it, it wasn't there.  If the vuln bundle is intended to be pushed to that version for a while, then I'd be willing to hard code 1.0.0.

Reply:
There is some bug: 1.0.0 appears to be an object, not a directory.

```
$ gsutil ls gs://definitions.stackrox.io/v4/vulnerability-bundles/
gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
...
gs://definitions.stackrox.io/v4/vulnerability-bundles/4.4.0/
gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/
```

Notice no `/` at the end.

```
$ gsutil ls -lh gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/  
164.02 MiB  2024-03-20T13:18:37Z  gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/vulns.json.zst
TOTAL: 1 objects, 171991321 bytes (164.02 MiB)
```

```
$ gsutil ls -lh gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
164.02 MiB  2024-03-20T13:18:40Z  gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0
TOTAL: 1 objects, 171991321 bytes (164.02 MiB)
```

Maybe those lines should end with trailing slash like
```
        gsutil cp "vulns/vulns.json.zst" "gs://definitions.stackrox.io/v4/vulnerability-bundles/dev/"
        gsutil cp "vulns/vulns.json.zst" "gs://definitions.stackrox.io/v4/vulnerability-bundles/1.0.0/"
```

Reply:
I think it should be good enough to hardcode `dev` and add a `# TODO` for now, if it actually works.

Reply:
Alright, I've made that change, but I haven't been able to figure out why the `release` gotag is not being used during compilation.  I created [ROX-23340](https://issues.redhat.com/browse/ROX-23340) as a follow up if you're okay to resolve this and merge.